04-cs-fundamentals/classroom/recursion.md

```diff
@@ -0,0 +1,343 @@
+    sum += num  # Add the current num to sum
+    num -= 1    # decrement num
+  end
+
+  return sum
+end
+```
+
+<details>
+  <summary>If you did this recursively, what is the base case?</summary>
+  Either num = 1 or num = 0.  In that case return num.
+</details>
+
+<details>
+  <summary>If you did this recursively, what do you do in the recursive case?</summary>
+  return num + natural_numbers_sum(num - 1)
+</details>
+
+**Write the recursive solution** 
+
+If you want the solution you can find it in [our examples folder](examples/natural_numbers_sum.rb).
+
+## Infinite Recursion
+
+What is wrong with this code:
+
+```ruby
+def fibonacci(num) 
+  return fibonacci(num - 1) + fibonacci(num - 2)
+end
+```
+
+<details>
+  <summary>What could be wrong with the above code?</summary>
+  This code produces a Stack Overflow Error because it performs infinite recursion.  The method needs a base case like return 1 if num <= 1
+</details>
+
+## Tail Recursion
+
+Look at the the code sample below.  How does it differ from the previous `natural_numbers_sum`.
+
+```ruby
+def natural_numbers_sum(num, sum = 0)
+  return sum if num == 0
+
+  return natural_numbers_sum(num - 1, sum + num)
+end
+```
+
+Go ahead and run the code.  It functions just the same, and for Ruby, the code, by default, functions the same.  On some systems however a compiler or interpreter can convert this into a loop, bypassing the time and space overhead of multiple function calls.
+
+This technique is called _Tail Recursion_.  In tail recursion, the recursive call is the **last** command executed in a method, by leaving nothing else in the calling method to do the interpreter can re-work the recursion into a loop.  Ruby, by default, [doesn't enable this feature](https://nithinbekal.com/posts/ruby-tco/) of tail recursion, but many other languages do. The Ruby language spec doesn't require the feature because it does not want to force every implementation to support it and tail recursion can make stack traces harder to understand since they transform method calls into loops with no cooresponding line numbers.
+
+In general a tail recursive method, like the example above carries the calculations in the parameters to avoid the calling method from needing to do any further calculations once the recursive call is complete.
+
+## Recursion vs Iteration, Which is Better
+
+Both loops and recursive solutions arrive at the same result, code gets repeated.  However recursive methods are often considered more _elegant_ because they can be written with less code.  Further some data structures like [binary search trees](../internship/binary-search-trees.md) are naturally recursive and recursive solutions are equally efficient and easier to write.  
+
+On the other hand, iterative solutions do not require use of the system stack, saving space, and the overhead of a method call (both a cost in space and time).  Thus often an iterative solution will be both faster and more space efficient.
```

Comment:
Maybe add something like: "Languages that use lots of recursive data structures use the aforementioned tail-cail-optimization to make up this gap in performance."